crates/ra_hir_def/src/item_scope.rs

```diff
@@ -126,17 +126,40 @@ impl ItemScope {
         let mut changed = false;
         let existing = self.visible.entry(name).or_default();
 
-        if existing.types.is_none() && def.types.is_some() {
-            existing.types = def.types;
-            changed = true;
+        match (existing.types, def.types) {
+            (None, Some(_)) => {
+                existing.types = def.types;
+                changed = true;
+            }
+            (Some(e), Some(d)) if e.0 != d.0 => {
+                existing.types = def.types;
```

Comment:
Let's use `mark::hit` here. Also would be nice to maybe abstract the three cases into a function/macro?

Reply:
Done. This should correspond with the `mark::check` in the test, right?

Reply:
Also I went with the macro because it's simpler to read. The function version requires generics and bounds because the macro tuple type doesn't match the other two cases.